### PR TITLE
[http] Add OAuth2 Token-Authentication with Token-Endpoint and grant-type password

### PR DESCRIPTION
**Pull request by BoBiene**
## OAuth2 Token-Authentication with Token-Endpoint and grant-type password

Add the possibility to request an OAuth Token (grant-type password).
The token is extracted from the response payload and added as Authorization Header to the requests.

When the token expires (configurable time) or a 401 is returned for a item-status-refresh the token gets invalidated and will be renewed on next request.

### Testing
baseURL: https://httpbin.org/bearer
authMode: TOKEN_PASSWORD
tokenEndpointURL: https://httpbin.org/post
tokenJsonPath: $

### CODE REVIEW

**Comment by BoBiene** [comment]
@J-N-K i don't know why the jenkins build is failing. Can you assist?

````log
[ERROR] Failed to execute goal org.openhab.tools.sat:sat-plugin:0.10.0:report (sat-all) on project org.openhab.binding.http:
[ERROR] Code Analysis Tool has found 1 error(s)!
[ERROR] Please fix the errors and rerun the build.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :org.openhab.binding.http
````

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpClientProvider.java [hunk 1]
```diff
@@ -28,12 +29,14 @@
      *
      * @return a HttpClient
      */
+    @NonNull
```
Comment (J-N-K) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpClientProvider.java:32]: The class is already @NonNullByDefault, why is @NonNull necessary here?

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/RateLimitedHttpClient.java [hunk 2]
```diff
@@ -112,18 +210,140 @@ private void stopProcessJob() {
 
     private void processQueue() {
         RequestQueueEntry queueEntry = requestQueue.poll();
-        if (queueEntry != null) {
-            queueEntry.future.complete(httpClient.newRequest(queueEntry.finalUrl));
+        execute(queueEntry);
```
Comment (J-N-K) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/RateLimitedHttpClient.java:213]: poll will return null if the queue is empty, resulting in an NPE

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/RateLimitedHttpClient.java [hunk 3]
```diff
@@ -112,18 +210,140 @@ private void stopProcessJob() {
...
+    }
+
+    private HttpClient getHttpClient() {
+        return Optional.ofNullable(this.httpClient)
```
Comment (J-N-K) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/RateLimitedHttpClient.java:217]: Creating an optional for null-checking is considered bad style.

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/RateLimitedHttpClient.java [hunk 4]
```diff
@@ -29,16 +47,24 @@
  */
 @NonNullByDefault
 public class RateLimitedHttpClient {
+    private final Logger logger = LoggerFactory.getLogger(RateLimitedHttpClient.class);
     private static final int MAX_QUEUE_SIZE = 1000; // maximum queue size
-    private HttpClient httpClient;
-    private int delay = 0; // in ms
+    private Thing thing;
```
Comment (J-N-K) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/RateLimitedHttpClient.java:52]: Please keep client and thing separated. Why did you remove the client from the constructor?

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/RateLimitedHttpClient.java [hunk 5]
```diff
@@ -112,18 +210,140 @@ private void stopProcessJob() {
...
+                        obj = ele.getAsJsonObject();
+                } else {
+                    throw new JsonSyntaxException("Given json is invalid");
+                }
+            }
+        }
+        return Optional.ofNullable(obj).orElseThrow(() -> new JsonSyntaxException("Given json is invalid"));
+    }
+
+    public class AuthenticatedRequestContext {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/RateLimitedHttpClient.java:305]: Please move to separate class if used by other classes

**Review by J-N-K — CHANGES_REQUESTED** [review]
Please also check for codestyle (SAT) warnings and compiler warnings.

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/RateLimitedHttpClient.java (hunk 3 above)
Comment (BoBiene) [reply bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/RateLimitedHttpClient.java:217]: Ok, i'm new to Java. How can i translate a @Nullable to be accepted by a @nonnull? 
this was the only way i found.

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/RateLimitedHttpClient.java (hunk 4 above)
Comment (BoBiene) [reply bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/RateLimitedHttpClient.java:52]: The intention was to keep the previous log entries and they got the Thing-id. 
Should i use a diffrent log entry or pass the thing id?

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/RateLimitedHttpClient.java (hunk 3 above)
Comment (J-N-K) [reply bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/RateLimitedHttpClient.java:217]: Why is this `@Nullable` in the first place? That's essentially the same question I had below: Why did you remove the client assignment from the constructor?

You could return the `Optional` and use `isPresent` or `ifPresent` in the caller or you could assign the field to a local variable and check that against null.

Comment (BoBiene) [reply bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/RateLimitedHttpClient.java:217]: I removed it from the ctor because in the execution flow it is never used. 
First use of the httpClient is after initialize is called.

So i thought it is better to instantiate in the initialize and not in the ctor to throw it away in the initialize.

**Comment by BoBiene** [comment]
@J-N-K how is the RateLimiting inteded to behave? Currently it is only "rate-limiting" upon item refresh and not for item-send commands. 
Should i fix this, and process the send command over the ratelimited client?

**Comment by Hilbrand** [comment]
Did you know openHAB has build in an OAuth2 client to handle the flow? I think if OAuth2 would be added it would make sense to use that library (for example used in Spotify binding). wdyt?

**Comment by J-N-K** [comment]
Correct, the send commands are currently not handled by the rate-Limited client. This was done intentionally to prevent long delays between sending a command on openHAB and the command to the device itself. My assumption was that sending commands is done less often than requesting states and therefore rate-limiting is not needed. Did you encounter problems with this approach?

regarding @Hilbrand‘s comment: I agree. If possible the implementation present in openHAB should be used.

**Comment by BoBiene** [comment]
> Did you know openHAB has build in an OAuth2 client to handle the flow? I think if OAuth2 would be added it would make sense to use that library (for example used in Spotify binding). wdyt?

@Hilbrand as far as i can see the Spotify-Binding uses the https://www.oauth.com/oauth2-servers/server-side-apps/ OAuth2 flow. What i implemented is more or less the Password Grand https://www.oauth.com/oauth2-servers/access-tokens/password-grant/ to be able to use a OAuth2 based REST API with Username & Password from OH.

**Comment by BoBiene** [comment]
> Did you know openHAB has build in an OAuth2 client to handle the flow? I think if OAuth2 would be added it would make sense to use that library (for example used in Spotify binding). wdyt?

@Hilbrand as far as i can see the Spotify-Binding uses the https://www.oauth.com/oauth2-servers/server-side-apps/ OAuth2 flow. What i implemented is more or less the Password Grand https://www.oauth.com/oauth2-servers/access-tokens/password-grant/

**Comment by BoBiene** [comment]
@J-N-K will it be OK with my implementation of the Password Grand?
Then i would implement your suggestions an update this pull request.

**Comment by J-N-K** [comment]
I am no longer maintaining this code, so I don‘t care.

**Review by J-N-K — APPROVED** [review]
